File: six/Browser/SearchEngine.swift
import Foundation

/// Where a query goes, and where the start page gets its suggestions. All four engines answer
/// suggestion requests in the same OpenSearch shape — `["query", ["suggestion", …]]` — so one
/// parser serves them all.
nonisolated enum SearchEngine: String, CaseIterable, Identifiable, Sendable {
    case duckDuckGo
    case google
    case bing
    case yandex

    var id: String { rawValue }

    var title: String {
        switch self {
        case .duckDuckGo: "DuckDuckGo"
        case .google: "Google"
        case .bing: "Bing"
        case .yandex: "Yandex"
        }
    }

    /// The engine in settings — one value for the address bar, every start page and the tools.
    @MainActor static var current: SearchEngine {
        get { SettingsStore.shared?.searchEngine ?? .duckDuckGo }
        set { SettingsStore.shared?.searchEngine = newValue }
    }

    /// What the engine calls the query in its own address. Three of them say `q`; Yandex says
    /// `text`, and both halves of this file — writing a search and reading one back — go through
    /// here rather than assuming.
    private var queryParameter: String {
        switch self {
        case .duckDuckGo, .google, .bing: "q"
        case .yandex: "text"
        }
    }

    func searchURL(for query: String) -> URL? {
        var components: URLComponents
        switch self {
        case .duckDuckGo: components = URLComponents(string: "https://duckduckgo.com/")!
        case .google: components = URLComponents(string: "https://www.google.com/search")!
        case .bing: components = URLComponents(string: "https://www.bing.com/search")!
        case .yandex: components = URLComponents(string: "https://yandex.ru/search/")!
        }
        components.queryItems = [URLQueryItem(name: queryParameter, value: query)]
        return components.url
    }

    /// The query behind one of this engine's results pages, if `url` is one — so history can show
    /// "паша · DuckDuckGo Search" instead of the results page's title.
    func query(from url: URL) -> String? {
        guard let host = url.host()?.lowercased(),
              let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.percentEncodedQueryItems,
              let raw = items.first(where: { $0.name == queryParameter })?.value else { return nil }
        let q = Self.decoded(raw)
        guard !q.isEmpty else { return nil }
        let bare = host.hasPrefix("www.") ? String(host.dropFirst(4)) : host
        switch self {
        case .duckDuckGo: return host == "duckduckgo.com" || host.hasSuffix(".duckduckgo.com") ? q : nil
        case .google: return host.hasPrefix("www.google.") && url.path() == "/search" ? q : nil
        case .bing: return (bare == "bing.com" || bare.hasSuffix(".bing.com")) && url.path() == "/search" ? q : nil
        // Yandex answers from a country domain per visitor — .ru, .com, .com.tr, .kz — and from
        // ya.ru; the results path is /search/, with /search/touch/ for the phone.
        case .yandex: return (bare == "ya.ru" || bare.hasPrefix("yandex.")) && url.path().hasPrefix("/search") ? q : nil
        }
    }

    /// A query string's space, however the page that wrote it spelled one.
    ///
    /// six writes `%20`; the engines write `+` in the address their own search box produces, and
    /// `URLComponents` decodes the escapes without touching the `+` — which is how history came to
    /// show "слово+раз". Substituting before the decode rather than after is what keeps a
    /// searched-for plus, which arrives as `%2B`, a plus.
    private static func decoded(_ percentEncoded: String) -> String {
        let spaced = percentEncoded.replacingOccurrences(of: "+", with: "%20")
        return spaced.removingPercentEncoding ?? percentEncoded
    }

    /// Whichever engine's results page this is.
    static func search(from url: URL) -> (engine: SearchEngine, query: String)? {
        for engine in allCases {
            if let query = engine.query(from: url) { return (engine, query) }
        }
        return nil
    }

    func suggestionsURL(for query: String) -> URL? {
        var components: URLComponents
        switch self {
        case .duckDuckGo:
            components = URLComponents(string: "https://duckduckgo.com/ac/")!
            components.queryItems = [URLQueryItem(name: "q", value: query), URLQueryItem(name: "type", value: "list")]
        case .google:
            components = URLComponents(string: "https://suggestqueries.google.com/complete/search")!
            components.queryItems = [URLQueryItem(name: "client", value: "firefox"), URLQueryItem(name: "q", value: query)]
        case .bing:
            components = URLComponents(string: "https://www.bing.com/osjson.aspx")!
            components.queryItems = [URLQueryItem(name: "query", value: query)]
        case .yandex:
            components = URLComponents(string: "https://suggest.yandex.ru/suggest-ff.cgi")!
            components.queryItems = [URLQueryItem(name: "part", value: query)]
        }
        return components.url
    }
}

// MARK: - Settings

/// The setting lives in the settings table; the knowledge of what its string means lives here,
/// beside the type it means it as. `SettingsStore` itself keeps only keys and strings.
extension SettingsStore {
    /// The engine six searches with. Read statically through `SearchEngine.current`.
    var searchEngine: SearchEngine {
        get { SearchEngine(rawValue: self[.searchEngine] ?? "") ?? .duckDuckGo }
        set { self[.searchEngine] = newValue.rawValue }
    }
}

File: android/core/src/main/kotlin/org/deffun/six/core/SearchEngine.kt
package org.deffun.six.core

import java.net.URI
import java.net.URLEncoder

/**
 * Where a query goes, and how the address field tells one from an address.
 *
 * [id] is the string the `settings` table holds, and it is the Swift enum's `rawValue` rather than
 * anything idiomatic here: a search engine chosen on the Mac and read on the phone has to be the
 * same engine, and the only thing carrying that across is this string.
 */
enum class SearchEngine(val id: String, val title: String) {
    DUCK_DUCK_GO("duckDuckGo", "DuckDuckGo"),
    GOOGLE("google", "Google"),
    ;

    fun searchUrl(query: String): String = when (this) {
        DUCK_DUCK_GO -> "https://duckduckgo.com/?q=${encode(query)}"
        GOOGLE -> "https://www.google.com/search?q=${encode(query)}"
    }

    /**
     * The query behind one of this engine's results pages, if that is what this is — so history can
     * show "пилав · DuckDuckGo Search" rather than the results page's own title.
     */
    fun query(url: String): String? {
        val uri = runCatching { URI(url) }.getOrNull() ?: return null
        val host = uri.host?.lowercase() ?: return null
        val q = queryParameter(uri.rawQuery, "q")?.takeIf { it.isNotEmpty() } ?: return null
        return when (this) {
            DUCK_DUCK_GO -> if (host == "duckduckgo.com" || host.endsWith(".duckduckgo.com")) q else null
            GOOGLE -> if (host.startsWith("www.google.") && uri.path == "/search") q else null
        }
    }

    companion object {
        /** What six searches with until the settings table says otherwise. */
        val DEFAULT = DUCK_DUCK_GO

        fun from(id: String?): SearchEngine = entries.firstOrNull { it.id == id } ?: DEFAULT

        /** Whichever engine's results page this is. */
        fun search(from: String): Pair<SearchEngine, String>? {
            for (engine in entries) {
                val query = engine.query(from) ?: continue
                return engine to query
            }
            return null
        }

        /**
         * Space is `%20`, not `+`.
         *
         * `URLEncoder` writes a form-encoded `+`, which every engine happens to accept and which
         * then does not match what `URLComponents` produces on the Mac. Two front ends writing the
         * same search as two different addresses is two rows in one history.
         */
        private fun encode(value: String): String =
            URLEncoder.encode(value, Charsets.UTF_8).replace("+", "%20")

        private fun decode(value: String): String =
            runCatching { java.net.URLDecoder.decode(value, Charsets.UTF_8) }.getOrDefault(value)

        private fun queryParameter(rawQuery: String?, name: String): String? =
            rawQuery
                ?.split('&')
                ?.firstNotNullOfOrNull { pair ->
                    val separator = pair.indexOf('=')
                    if (separator <= 0 || pair.substring(0, separator) != name) null
                    else decode(pair.substring(separator + 1))
                }
    }
}

/**
 * Address-bar input, turned into something to load.
 *
 * The rule is the Mac's, in `URL.fromUserInput`: a recognised scheme is an address, a scheme-less
 * thing that could be a host gets `https://`, and everything else is a search. It is small enough to
 * look obvious and specific enough that reimplementing it by eye on a second platform would produce
 * a browser where the same typing does different things.
 */
object UserInput {

    private val ADDRESS_SCHEMES = setOf("http", "https", "file", "about")

    /** What to load for what was typed, or null if nothing was. */
    fun url(raw: String, engine: SearchEngine = SearchEngine.DEFAULT): String? {
        val text = raw.trim()
        if (text.isEmpty()) return null

        val scheme = runCatching { URI(text).scheme }.getOrNull()?.lowercase()
        if (scheme in ADDRESS_SCHEMES) return text

        if (looksLikeHost(text)) return "https://$text"

        return engine.searchUrl(text)
    }

    /** Does this look like something to open rather than something to search for? */
    fun looksLikeAddress(raw: String): Boolean {
        val text = raw.trim()
        if (text.isEmpty() || text.contains(' ')) return false
        val scheme = runCatching { URI(text).scheme }.getOrNull()?.lowercase()
        if (scheme in ADDRESS_SCHEMES) return true
        return text.contains('.') || text.startsWith("localhost")
    }

    private fun looksLikeHost(text: String): Boolean =
        !text.contains(' ') && (text.contains('.') || text.startsWith("localhost"))
}

// MARK: - The setting

/**
 * The setting lives in the `settings` table; the knowledge of what its string means lives here,
 * beside the type it means it as — the same split the Mac makes, and the reason [SettingsStore]
 * itself knows only keys and strings.
 *
 * `search.engine` is that table's key, and it is the Mac's. A key invented here would be a setting
 * the other platform never sees change.
 */
var SettingsStore.searchEngine: SearchEngine
    get() = SearchEngine.from(this[SettingsKeys.SEARCH_ENGINE])
    set(value) {
        this[SettingsKeys.SEARCH_ENGINE] = value.id
    }

/** Keys in the shared `settings` table. Only the ones this platform reads or writes. */
object SettingsKeys {
    const val SEARCH_ENGINE = "search.engine"
    const val CENTERS_FOCUS = "layout.centersFocus"
    const val COLUMN_WIDTH = "layout.columnWidth"
    const val DEFAULT_PROFILE = "profile.default"
}
